orttraining/orttraining/core/agent/training_agent.cc

```diff
@@ -2,135 +2,67 @@
-    bg_threads_[run_id] = std::move(bg_thread);
+    ORT_ENFORCE(utils::InitializeFeedFetchCopyInfo(session_state, *fw_feeds_fetches_manager_) == Status::OK());
   }
-
-  onnxruntime::contrib::OrtTasks::GetInstance().CreateBackgroundTask(run_id);
-
-  LOGS(*inference_session_.GetLogger(), VERBOSE) << "InferenceSession::Forward() call created a task with run_id " << run_id;
-
-  // background task is setup, unblock background thread to continue
-  setup_promise.set_value();
-
-  // Wait for data/signal from
-  // 1. Yield op, if the bg thread sucessfully reached Yield's signal point
-  // 2. The end of bg thread, if it hit execptions and returned earlier
-  auto forward_outputs = onnxruntime::contrib::OrtTasks::GetInstance().WaitForForwardOutputs(run_id);
-  const Status& forward_status = forward_outputs.first;
-  user_outputs = std::move(forward_outputs.second);
-
-  // background thread has completed without hitting Yield Op
-  if (!forward_status.IsOK()) {
-    std::thread thread;
-    {
-      std::lock_guard<std::mutex> lock(bg_threads_mutex_);
-      std::swap(thread, bg_threads_[run_id]);
-      bg_threads_.erase(run_id);
+  {
+    FeedsFetchesManager::Create(bw_feed_names, bw_fetches_names, session_state.GetOrtValueNameIdxMap(), bw_feeds_fetches_manager_);
+    auto& fetch_info = bw_feeds_fetches_manager_->GetMutableFetchesDeviceCopyInfo();
+    for (size_t i = 0, end = bw_fetches_names.size(); i < end; ++i) {
+      fetch_info[i].target_device = bw_outputs_device_info[i];
     }
-    ORT_ENFORCE(thread.joinable());
-    thread.join();
-    onnxruntime::contrib::OrtTasks::GetInstance().RemoveTask(run_id);
-    return forward_status;
+
+    ORT_ENFORCE(utils::InitializeFeedFetchCopyInfo(session_state, *bw_feeds_fetches_manager_) == Status::OK());
   }
 
-  return Status::OK();
+  auto bp = inference_session_.GetBreakpointAndEndPoint();
+  fw_program_counter_end_ = bp.first - 1;
+  bw_program_counter_end_ = bp.second;
 }
 
-common::Status TrainingAgent::RunBackward(int64_t run_id, const std::vector<OrtValue>& backward_output_grads) {
-  LOGS(*inference_session_.GetLogger(), VERBOSE) << "Running TrainingAgent::Backward() with run_id " << run_id;
+TrainingAgent::~TrainingAgent(){};
 
-  // resume background thread
-  onnxruntime::contrib::OrtTasks::GetInstance().SetBackwardInputs(run_id, backward_output_grads, false);
-
-  Status bg_thread_status = onnxruntime::contrib::OrtTasks::GetInstance().WaitForStatus(run_id);
-
-  std::thread bg_thread;
-  {
-    std::lock_guard<std::mutex> lock(bg_threads_mutex_);
-    std::swap(bg_thread, bg_threads_[run_id]);
-    bg_threads_.erase(run_id);
-  }
-
-  // wait for bg_thread to complete
-  ORT_ENFORCE(bg_thread.joinable());
-  bg_thread.join();
-  onnxruntime::contrib::OrtTasks::GetInstance().RemoveTask(run_id);
-
-  return bg_thread_status;
+void TrainingAgent::RunForward(std::vector<OrtValue>& feeds, std::vector<OrtValue>& fetches, PartialGraphExecutionState& state) {
```

Comment:
Nit: Would it make sense to make `feeds` a `const std::vector<OrtValue>&`? (I'm not sure if this would enable pybind to perform any more optimizations... probably not with the opaque type fix, but beforehand it probably had to do a copy-out of all the unchanged feed tensors....)

Reply:
const makes sense regardless.

---
In reply to: [613571907](https://github.com/microsoft/onnxruntime/pull/7324#discussion_r613571907) [](ancestors = 613571907)